modules/call-dispatch/src/lib.rs

```diff
@@ -194,11 +217,21 @@ impl<T: Trait<I>, I: Instance> MessageDispatch<T::MessageId> for Module<T, I> {
 
 		// prepare dispatch origin
 		let origin_account = match message.origin {
-			CallOrigin::BridgeAccount => bridge_account_id(bridge, CALL_DISPATCH_MODULE_PREFIX),
-			CallOrigin::RealAccount(source_public, target_public, target_signature) => {
+			CallOrigin::SourceRoot => {
+				let encoded_id = derive_account_id::<T::SourceChainAccountId>(bridge, SourceAccount::Root);
+				match T::AccountId::decode(&mut &encoded_id[..]) {
+					Ok(id) => id,
+					Err(_) => {
+						let error = <Error<T, I>>::InvalidAccountId.into();
+						Self::deposit_event(RawEvent::InvalidAccountId(bridge, id, error));
+						return;
```

Comment:
Tbh I'm not sure how to test the codepath for this right now. The test runtime has an `AccountId` of `u64` which I don't think can ever fail to be decoded. I _could_ introduce a wrapper type and then have a sketchy `Decode` implementation which fails on a specific input, but I don't like that.

Reply:
Could we use `UncheckedFrom<H256>` or `From<H256>` or even `TryFrom<H256>` instead of `Decode` to get `AccountId`? I'm not convinced that `Decode` is the right choice here at all, cause usually SCALE can decode something even from garbage data. (like decoding `u64` from `H256`).

Alternatively, let's make the conversion explicit by using `Convert<H256, AccountId>` trait from substrate and passing an extra parameter, so that the runtime can provide a proof that the conversion is always successful.